### PR TITLE
fix(synthetic): consider ancestors when determining click-focusability

### LINKED ISSUE #1382: [delegatesFocus] Focus should transfer to focusable ancestor (in the same shadow) when click target is not focusable
### Description

When the user clicks on a focusable element, the delegatesFocus polyfill disables all the magic and lets the browser do its thing. This implementation breaks down when the click target is a non-focusable element wrapped by a focusable element.

#### Expected Behavior

Focus should apply to the focusable element wrapping the non-focusable element when the non-focusable element is clicked.

#### Observed Behavior

Clicks on non-focusable elements causes focus to be forwarded to the next focusable element outside of the shadow, even the the click target is wrapped by a focusable element (such as a `button`).

It is as though the user arrived at the target via sequential focus navigation, sees that the shadow host has a tabindex value of -1, and forwards focus to the next focusable element outside of the shadow. 

#### Steps to Reproduce

https://playground.lwcjs.org/projects/CsB3R9LDu/6/edit

### PR DESCRIPTION
**Pull request by ekashida**
## Details

Fixes #1382 

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅

### CODE REVIEW

File: packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts [hunk 1]
```diff
@@ -104,14 +105,23 @@ const focusableTagNames = {
 // This function based on https://allyjs.io/data-tables/focusable.html
 // It won't catch everything, but should be good enough
 // There are a lot of edge cases here that we can't realistically handle
-function isFocusable(element: HTMLElement): boolean {
-    const tagName = tagNameGetter.call(element);
-    return (
+function isFocusable(element: HTMLElement, hostElement: HTMLElement): boolean {
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts]: `hostElement` is probably `Element` now that we have adjusted that... same for element

Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts]: maybe not... if this is only for custom elements, then hostElement is HTMLElement

File: packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts [hunk 2]
```diff
@@ -104,14 +105,23 @@ const focusableTagNames = {
...
             hasAttribute.call(element, 'contenteditable') ||
-            hasOwnProperty.call(focusableTagNames, tagName))
-    );
+            hasOwnProperty.call(focusableTagNames, tagNameGetter.call(element)));
+
+    if (isTrue(focusable)) {
+        return true;
+    }
+
+    const nextElement =
```
Comment (caridy) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts]: I'm not sure about this condition... I don't understand it.

`element.parentNode === element.getRootNode()` means that you're a top level element of the shadow. Maybe the name `nextElement` is what confuses me.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts (hunk 1 above)
Comment (ekashida) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts]: `hostElement` is always a custom element and `element` is every element in the ancestor chain (composed) between the `target` and the `hostElement`.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts (hunk 2 above)
Comment (ekashida) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/focus.ts]: We want to keep bubbling up through any shadow boundaries until we hit the custom element of the component that is delegating focus. I can change that check to something else if there's a better way to detect a shadow root.

**Pull request merged**